src/core/handlers/RequestHandler.ts

```diff
@@ -57,6 +56,11 @@ export type AsyncResponseResolverReturnType<
       MaybeAsyncResponseResolverReturnType<ResponseBodyType>,
       MaybeAsyncResponseResolverReturnType<ResponseBodyType>
     >
+  | AsyncGenerator<
```

Comment:
Older versions of TypeScript (4.7 - 5.2) have trouble digesting this type:

```
 FAIL  resolver-generator.test-d.ts > supports async generator function as response resolver
TypeCheckError: Argument of type '() => AsyncGenerator<StrictResponse<JsonBodyType>, StrictResponse<JsonBodyType>, any>' is not assignable to parameter of type 'HttpResponseResolver<never, never, { value string; }>'.
  Type 'AsyncGenerator<StrictResponse<JsonBodyType>, StrictResponse<JsonBodyType>, any>' is not assignable to type 'AsyncResponseResolverReturnType<{ value string; }>'.
    Type 'AsyncGenerator<StrictResponse<JsonBodyType>, StrictResponse<JsonBodyType>, any>' is not assignable to type 'AsyncGenerator<ResponseResolverReturnType<{ value string; }>, ResponseResolverReturnType<{ value string; }>, ResponseResolverReturnType<{ value string; }>>'.
      The types returned by 'next(...)' are incompatible between these types.
        Type 'Promise<IteratorResult<StrictResponse<JsonBodyType>, StrictResponse<JsonBodyType>>>' is not assignable to type 'Promise<IteratorResult<ResponseResolverReturnType<{ value string; }>, ResponseResolverReturnType<{ value string; }>>>'.
          Type 'IteratorResult<StrictResponse<JsonBodyType>, StrictResponse<JsonBodyType>>' is not assignable to type 'IteratorResult<ResponseResolverReturnType<{ value string; }>, ResponseResolverReturnType<{ value string; }>>'.
            Type 'IteratorYieldResult<StrictResponse<JsonBodyType>>' is not assignable to type 'IteratorResult<ResponseResolverReturnType<{ value string; }>, ResponseResolverReturnType<{ value string; }>>'.
              Type 'IteratorYieldResult<StrictResponse<JsonBodyType>>' is not assignable to type 'IteratorYieldResult<ResponseResolverReturnType<{ value string; }>>'.
                Type 'StrictResponse<JsonBodyType>' is not assignable to type 'ResponseResolverReturnType<{ value string; }>'.
```

## Reproduction steps

- TypeScript 4.7
- `pnpm test:ts`

Reply:
That is not exactly minimal ;p it still depends on everything in the MSW's codebase. I could really use a self-contained TS playground here.

Reply:
I stumbled on this and tried to bisect (via https://www.npmjs.com/package/every-ts) when this would have changed between 5.2 and 5.3.

What I discovered is that it's the literal version bump from 5.2 to 5.3; something in this package or its dependencies must be doing some typesVersions shenanigans that is causing this error.

Reply:
Hmmm, no, I just flipped back and forth between the commits and now the 5.3 version change is failing. This makes no sense.

Reply:
Skipping using `test:ts` and just doing regular compilation shows that https://github.com/microsoft/TypeScript/pull/51196 made this code legal. I'm not sure that there's really anything that can be done if you're hitting a limitation that was fixed.

Reply:
Thank you so much for investigating this, @jakebailey! 

I'm glad to hear the team has fixed that (what a specific issue though!). Sadly, I don't expect they will backport the fix to all downstream versions that are not EOL. As CI reports, TS v5.3+ is fine with this change, but there's a huge chunk of versions MSW still supports before that. 

But one thing I'd like to clarify from your findings: is that a fundamentally TypeScript issue? Can we overcome it in any way on our end? Because if not, then this issue, transitively, also becomes a TS-caused issue, and we can only fix it once v5.3 becomes the minimal supported one.

Reply:
We in general don't backport anything except critical regression fixes to the recent version (i.e. we're patching 5.5 right now).

I'm honestly not an expert in these types or their intracacies, so I can't really offer any other solution up for this... At best I could play around with the PR to attempt to make something pass, though I think anyone could probably do just as good as me in that area 😅

Reply:
If you have the time, I'd be grateful! I gave it my best but my knowledge here ran short.

Reply:
> We in general don't backport anything except critical regression fixes to the recent version (i.e. we're patching 5.5 right now).
> 
> I'm honestly not an expert in these types or their intracacies, so I can't really offer any other solution up for this... At best I could play around with the PR to attempt to make something pass, though I think anyone could probably do just as good as me in that area 😅

I was looking into this a bit too - happy to have assistance if you have ideas - it's been an interesting one, and also not super familiar with these types

Reply:
I got some time to take a look and believe I have a solution which actually improves things further.

Firstly, the code actually only depends on the existence of `Symbol.iterator` (or in this PR, `Symbol.asyncIterator`). This means that it's not actually working with generators at all, it's working with `Iterable`/`AsyncIterables` (and in fact shows that `isIterable` is incorrectly defined).

By switching the types over to use types defined with _just_ those two props, inference works again.

Fixing the implementation to reflect those types takes a little bit more code, but the upshot is that the final result allows users to pass in any `Iterable`/`AsyncIterable` they want, not just generators. I think this would actually allow people to pass in lists/sets of responses if they wanted, though I don't know if that's desirable or not.

How can I send this change to you? This is just a branch on this repo, so should I just PR it over? Here's a patch you could apply as well: https://gist.github.com/jakebailey/9702b2e0f16e2c9ce59799e40aefc7e7

Reply:
@jakebailey, that is incredible! 😮 I would love to give this a try. 

Please, could you open a PR to _this feature branch_ (`fix/generator-resolver`)? Let me know if you need any help. 

What you are proposing sounds fantastic. I think that's the way to go. Need to see whether it complies with the other expectations toward the resolvers (especially the `isUsed` stuff).

Reply:
Yes, I can do that.

Reply:
FWIW I did compile and test locally, and it all seemed to work, though I did not acutally test the full range of TS versions (but I can't imagine it wouldn't work, maybe I'm jinxing myself)

Reply:
Happy to give this a try once it's here 👀 Installing different `typescript` dependency versions triggers the `test:ts` into using that version for testing. The CI will also cover all the ranges for us. So excited we may actually ship this!

Reply:
Sent https://github.com/mswjs/msw/pull/2213.

FWIW if you're looking for multi-version testing and aren't too tied to vitest, https://github.com/tstyche/tstyche is one option I've heard of which automatically runs multiple versions of TS without having to install them explicitly.

Reply:
That's nice! Something to consider in the setup.